aten/src/ATen/native/cpu/int4mm_kernel.cpp

```diff
@@ -0,0 +1,597 @@
+//   weight:
+//     [Lane0] N0...15:  { a0,  a1,  a2, ...}
+//     [Lane1] N16...32: {a16, a17, a18, ...}
+//
+//  packed:
+//    [Lane01] N0...32: {a16|a0, a17|a1, a18|a2, ...}
+//
+//  Note: When N is 16, pack with 64-bit format
+//
+// 3 non-vectorized packed format:
+//   Do 64-bit unpacking at a time.
+//
+//   weight: {a0, a1, a2, a3, ..., a14, a15}
+//   packed: {a1|a0, a3, a2, ..., a15|a14}
+//
+void weight_to_int4pack_kernel(
+    const Tensor& weight_packed,
+    const Tensor& weight,
+    int N, int K) {
+
+  auto weight_packed_data = reinterpret_cast<uint8_t*>(weight_packed.data_ptr());
+  const auto weight_data = weight.data_ptr<int32_t>();
+
+  // 64 for avx512 and 64 for avx2/non-vectorized
+  constexpr int BLOCK_N = vec::Vectorized<float>::size() * 4;
+  const int NB =  (N + BLOCK_N - 1) / BLOCK_N;
+
+  // parallel on NB blocks
+  at::parallel_for(0, NB, 0, [&](int begin, int end) {
+    for (const auto i : c10::irange(begin, end)) {
+      int nb_size = std::min(BLOCK_N, N - i * BLOCK_N);
+
+      const int32_t* src = weight_data + i * BLOCK_N * K;
+      uint8_t* dst = weight_packed_data + i * K * BLOCK_N / 2;
+      for (const auto k : c10::irange(K)) {
+#if defined(CPU_CAPABILITY_AVX512) && !defined(_MSC_VER)
+        if (nb_size == BLOCK_N) {
+          for (const auto d : c10::irange(16)) {
+            int32_t val0 = src[(d +  0) * K + k];
+            int32_t val1 = src[(d + 16) * K + k];
+            int32_t val2 = src[(d + 32) * K + k];
+            int32_t val3 = src[(d + 48) * K + k];
+
+            uint8_t packed02 = (((uint8_t)(val2) << 4)) | ((uint8_t)(val0));
+            uint8_t packed13 = (((uint8_t)(val3) << 4)) | ((uint8_t)(val1));
+
+            dst[k * 32 + d] = packed02;
+            dst[k * 32 + 16 + d] = packed13;
+          }
+        } else {
+          // for nb_size 16, 32, 48
+          for (int n = 0; n < nb_size; n += 2) {
+            int32_t val0 = src[n * K + k];
+            int32_t val1 = src[n * K + K + k];
+
+            uint8_t packed = (((uint8_t)(val1) << 4)) | ((uint8_t)(val0));
+            dst[k * nb_size / 2 + n / 2] = packed;
+          }
+        }
+#elif defined(CPU_CAPABILITY_AVX2) && !defined(_MSC_VER)
+        if (nb_size == BLOCK_N) {
+          // for nb_size 32
+          for (const auto d : c10::irange(16)) {
+            int32_t val0 = src[(d + 0) * K + k];
+            int32_t val1 = src[(d + 16) * K + k];
+
+            uint8_t packed01 = (((uint8_t)(val1) << 4)) | ((uint8_t)(val0));
+            dst[k * 16 + d] = packed01;
+          }
+        } else {
+          // for nb_size 16
+          for (int n = 0; n < nb_size; n += 2) {
+            int32_t val0 = src[n * K + k];
+            int32_t val1 = src[n * K + K + k];
+
+            uint8_t packed = (((uint8_t)(val1) << 4)) | ((uint8_t)(val0));
+            dst[k * nb_size / 2 + n / 2] = packed;
+          }
+        }
+#else
+        for (int n = 0; n < nb_size; n += 2) {
+          int32_t val0 = src[n * K + k];
```

Comment:
I'm curious, what is the reason for transposing the tensor while packing it?

Reply:
The major reason is that x86 doesn't have a horizontal reduce from a SIMD vector to a Scalar. Note that on avx512, `_mm512_reduce_add_ps` is a SEQUENCE instruction, which means it will be translated into multiple uops (6 or 7 I don't remember precisely).

If we do gemm in the form of NT (which is default in pytorch Linear definition): {M, K} * {N, K}, we will end up doing dot-product on each m row by n col, and eventually we have a horizontal reduce and write to C would be scalar.

A better approach would be NN： {M, K} * {K, N},  we are doing FMA, broadcast each A[row_index] to a vector and load 4 or 6 vectors from B cols, then use C as an accumulator, we won't need a horizontal reduce and write to C would be a vectorized store.

Above is the explanation of avx512f with fp32 FMA, the other ISAs are similar:
* For avx512f, we transpose B from [N, K] to [K, N]
* For avx-vnni, we can assume that 4 * int8 is a float32, so we packed B from [N, K] to [K/4, N, 4], e.g. [K/4, N4]
* For avx512-bf16 or avx512-f16, we can assume that 2 * bf16 or 2 * f16 is a float32, so we packed B from [N, K] to [K/2, N, 2], e.g. [K/2, N2]
* For amx-vnni, it is just doing 16 cols together, you may assume that [K/4, N4] = [K/4, 64]
* for amx-bf16 or amx-f16, also it it doing 16 cols together, you may assume that [K/2, N2] = [k/2, 32]

So you see, all the gemms (fp32, int8, bf16) are following the same pattern.

The above is the explanation of the inner block, usually we need to break the weight B into multiples of inner blocks, this is ISA dependent:

For this piece of code, on avx512 we have 32 regs and we want to use them efficiently. A will take 1 reg, B will take 4, and C will take 4 by 4, the zero and scale takes 2 by 4. So we end up having block size of 4 x 64 (4 x 4 x 16 floats). avx512-vnni, avx512-bf16 follow the same rule. avx2 would be different as it has only 16 regs.

For amx, we have 8 tiles. Usually, the pattern is 2-2-4 or 1-3-3, which means A takes 2 tiles and B takes 2 tiles, C takes 4 tiles. and each tile computes 16 rows and 16 cols, so the inner block size is usually 16x16 and we handle 2 A blocks and 2 B blocks at a tiles (the 2-2-4 pattern). When M < 16, we use 1-3-3 pattern, waste 1 tile here.

It is often very difficult to understand the packed layout if we read from oneDNN primitive verbose. We can break it down into 2 parts and in this way it would be much easier to sort out the layout:
* the inner block (described in the front), are all the same no matter int8, bf16, f16, f32;
* the outer block (described in the latter), are related to register allocation and caching behaviors.